--- a/src/namestore/gnunet-namestore-gtk.c
+++ b/src/namestore/gnunet-namestore-gtk.c
@@ -266,7 +266,7 @@
   /**
    * Private key of target zone.
    */
-  struct GNUNET_CRYPTO_PrivateKey pk;
+  struct GNUNET_CRYPTO_BlindablePrivateKey pk;
 
   /**
    * Data to free.
@@ -413,12 +413,12 @@
 /**
  * Private key of the zone we are currently editing.
  */
-static const struct GNUNET_CRYPTO_PrivateKey *pkey;
+static const struct GNUNET_CRYPTO_BlindablePrivateKey *pkey;
 
 /**
  * Public key of the zone we are currently editing.
  */
-static struct GNUNET_CRYPTO_PublicKey pubkey;
+static struct GNUNET_CRYPTO_BlindablePublicKey pubkey;
 
 /**
  * Pseudonym of the current zone we are editing.
@@ -948,7 +948,7 @@
  */
 static void
 handle_records_for_merge (void *cls,
-                          const struct GNUNET_CRYPTO_PrivateKey *zone,
+                          const struct GNUNET_CRYPTO_BlindablePrivateKey *zone,
                           const char *label,
                           unsigned int rd_count,
                           const struct GNUNET_GNSRECORD_Data *rd)
@@ -1149,7 +1149,7 @@
     }
     else
     {
-      const struct GNUNET_CRYPTO_PrivateKey *pk;
+      const struct GNUNET_CRYPTO_BlindablePrivateKey *pk;
       struct MoveOperationContext *moc;
       struct GNUNET_HashCode query;
 
@@ -1733,7 +1733,13 @@
   GNUNET_asprintf (&edc->liblow,
                    "libgnunet_plugin_gtk_namestore_%s",
                    typename);
-  GNUNET_STRINGS_utf8_tolower (edc->liblow, edc->liblow);
+  {
+    char *liblow;
+
+    liblow = GNUNET_STRINGS_utf8_tolower (edc->liblow);
+    GNUNET_free (edc->liblow);
+    edc->liblow = liblow;
+  }
   edc->plugin = GNUNET_PLUGIN_load (GNUNET_GTK_project_data (),
                                     edc->liblow,
                                     edc);
@@ -2424,7 +2430,7 @@
  */
 static void
 zone_iteration_proc (void *cls,
-                     const struct GNUNET_CRYPTO_PrivateKey *zone_key,
+                     const struct GNUNET_CRYPTO_BlindablePrivateKey *zone_key,
                      const char *name,
                      unsigned int rd_count,
                      const struct GNUNET_GNSRECORD_Data *rd)
@@ -2782,7 +2788,7 @@
  */
 static void
 iop_create_cont (void *cls,
-                 const struct GNUNET_CRYPTO_PrivateKey *priv,
+                 const struct GNUNET_CRYPTO_BlindablePrivateKey *priv,
                  enum GNUNET_ErrorCode ec)
 {
   GtkEntry *entry;
--- a/src/namestore/plugin_gtk_namestore_phone.c
+++ b/src/namestore/plugin_gtk_namestore_phone.c
@@ -124,7 +124,7 @@
 {
   GtkEditable *entry;
   const gchar *preedit;
-  struct GNUNET_CRYPTO_PublicKey pub;
+  struct GNUNET_CRYPTO_BlindablePublicKey pub;
 
   entry = GTK_EDITABLE (
     gtk_builder_get_object (builder, "edit_dialog_phone_peer_entry"));
--- a/src/namestore/plugin_gtk_namestore_pkey.c
+++ b/src/namestore/plugin_gtk_namestore_pkey.c
@@ -99,7 +99,7 @@
 {
   GtkEditable *entry;
   const gchar *preedit;
-  struct GNUNET_CRYPTO_PublicKey pub;
+  struct GNUNET_CRYPTO_BlindablePublicKey pub;
 
   entry =
     GTK_EDITABLE (gtk_builder_get_object (builder, "edit_dialog_pkey_entry")),
